File: go-sdk/sdk/object-store/object_store.go
package objectstore

import (
	"fmt"
	regexp2 "regexp"

	"github.com/go-logr/logr"
	"github.com/konstellation-io/kre-runners/go-sdk/v1/internal/errors"
	"github.com/nats-io/nats.go"
	"github.com/spf13/viper"
)

type ObjectStore struct {
	logger          logr.Logger
	objStore        nats.ObjectStore
	objectStoreName string
}

func NewObjectStore(logger logr.Logger, jetstream nats.JetStreamContext) (*ObjectStore, error) {
	objectStoreName := viper.GetString("nats.object-store")

	logger = logger.WithName("[OBJECT STORE]")

	objStore, err := initObjectStoreDeps(logger, jetstream, objectStoreName)
	if err != nil {
		return nil, err
	}

	return &ObjectStore{
		logger:          logger,
		objStore:        objStore,
		objectStoreName: objectStoreName,
	}, nil
}

func initObjectStoreDeps(logger logr.Logger, jetstream nats.JetStreamContext,
	objectStoreName string) (nats.ObjectStore, error) {
	if objectStoreName != "" {
		objStore, err := jetstream.ObjectStore(objectStoreName)
		if err != nil {
			return nil, fmt.Errorf("error initializing object store: %w", err)
		}

		return objStore, nil
	}

	logger.Info("Object store not defined. Skipping object store initialization.")
	return nil, nil
}

func (ob ObjectStore) List(regexp ...string) ([]string, error) {
	if ob.objStore == nil {
		return nil, errors.ErrUndefinedObjectStore
	}

	objStoreList, err := ob.objStore.List()
	if err != nil {
		return nil, fmt.Errorf("error listing objects from the object store: %w", err)
	}

	var pattern *regexp2.Regexp

	if len(regexp) > 0 && regexp[0] != "" {
		pat, err := regexp2.Compile(regexp[0])
		if err != nil {
			return nil, fmt.Errorf("error compiling regexp: %w", err)
		}

		pattern = pat
	}

	var response []string

	for _, objName := range objStoreList {
		if pattern == nil || pattern.MatchString(objName.Name) {
			response = append(response, objName.Name)
		}
	}

	//nolint: gomnd
	ob.logger.V(2).Info("Files successfully listed", "Object store", ob.objectStoreName)

	return response, nil
}

func (ob ObjectStore) Get(key string) ([]byte, error) {
	if ob.objStore == nil {
		return nil, errors.ErrUndefinedObjectStore
	}

	response, err := ob.objStore.GetBytes(key)
	if err != nil {
		return nil, fmt.Errorf("error retrieving object with key %s from the object store: %w", key, err)
	}

	//nolint: gomnd
	ob.logger.V(1).Info("File successfully retrieved from object store", key, ob.objectStoreName)

	return response, nil
}

func (ob ObjectStore) Save(key string, payload []byte) error {
	if ob.objStore == nil {
		return errors.ErrUndefinedObjectStore
	}

	if payload == nil {
		return errors.ErrEmptyPayload
	}

	_, err := ob.objStore.PutBytes(key, payload)
	if err != nil {
		return fmt.Errorf("error storing object to the object store: %w", err)
	}

	//nolint: gomnd
	ob.logger.V(1).Info("File successfully stored in object store", key, ob.objectStoreName)

	return nil
}

func (ob ObjectStore) Delete(key string) error {
	if ob.objStore == nil {
		return errors.ErrUndefinedObjectStore
	}

	err := ob.objStore.Delete(key)
	if err != nil {
		return fmt.Errorf("error retrieving object with key %s from the object store: %w", key, err)
	}

	//nolint: gomnd
	ob.logger.V(1).Info("File successfully deleted in object store", key, ob.objectStoreName)

	return nil
}

func (ob ObjectStore) Purge(regexp ...string) error {
	if ob.objStore == nil {
		return errors.ErrUndefinedObjectStore
	}

	var pattern *regexp2.Regexp

	if len(regexp) > 0 && regexp[0] != "" {
		pat, err := regexp2.Compile(regexp[0])
		if err != nil {
			return fmt.Errorf("error compiling regexp: %w", err)
		}

		pattern = pat
	}

	objects, err := ob.List()
	if err != nil {
		return fmt.Errorf("error listing objects from the object store: %w", err)
	}

	for _, objectName := range objects {
		if pattern == nil || pattern.MatchString(objectName) {
			ob.logger.V(1).Info("Deleting object", "Object key", objectName) //nolint: gomnd

			err := ob.objStore.Delete(objectName)
			if err != nil {
				return fmt.Errorf("error purging objects from the object store: %w", err)
			}
		}
	}

	//nolint: gomnd
	ob.logger.V(2).Info("Files successfully purged", "Object Store", ob.objectStoreName)

	return nil
}

File: py-sdk/sdk/sdk/object_store/object_store.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Optional

import loguru
from loguru import logger
from nats.js.client import JetStreamContext
from nats.js.errors import NotFoundError, ObjectNotFoundError
from nats.js.object_store import ObjectStore as NatsObjectStore
from vyper import v

from sdk.object_store.exceptions import (
    FailedCompilingRegexpError,
    FailedDeletingFileError,
    FailedGettingFileError,
    FailedListingFilesError,
    FailedObjectStoreInitializationError,
    FailedPurgingFilesError,
    FailedSavingFileError,
    UndefinedObjectStoreError,
)

UNDEFINED_OBJECT_STORE = "object store not defined"


@dataclass
class ObjectStore:
    js: JetStreamContext
    object_store_name: Optional[str] = None
    object_store: Optional[NatsObjectStore] = None
    logger: loguru.Logger = logger.bind(context="[OBJECT STORE]")

    def __post_init__(self) -> None:
        self.object_store_name = v.get("nats.object_store")
        if self.object_store_name:
            self.logger = logger.bind(context=f"[OBJECT STORE: {self.object_store_name}]")

    async def initialize(self) -> None:
        if self.object_store_name:
            object_store = await self._init_object_store()
            self.object_store = object_store
        else:
            self.logger.info("object store not defined [skipped]")

    async def _init_object_store(self) -> NatsObjectStore:
        try:
            assert isinstance(self.object_store_name, str)
            object_store = await self.js.object_store(self.object_store_name)
            self.logger.debug(f"object store {self.object_store_name} successfully initialized")
            return object_store
        except Exception as e:
            self.logger.warning(f"failed initializing object store {self.object_store_name}: {e}")
            raise FailedObjectStoreInitializationError(error=e)

    async def list(self, regexp: Optional[str] = None) -> list[str]:
        if not self.object_store:
            self.logger.warning(UNDEFINED_OBJECT_STORE)
            raise UndefinedObjectStoreError

        try:
            objects = await self.object_store.list(ignore_deletes=True)
        except NotFoundError as e:
            self.logger.debug(f"no files found in object store {self.object_store_name}: {e}")
            return []
        except Exception as e:
            self.logger.warning(f"failed listing files from object store {self.object_store_name}: {e}")
            raise FailedListingFilesError(error=e)

        pattern = None
        if regexp:
            try:
                pattern = re.compile(regexp)
            except Exception as e:
                self.logger.warning(f"failed compiling regexp {regexp}: {e}")
                raise FailedCompilingRegexpError(error=e)

        response = []
        for obj in objects:
            obj_name = obj.name
            if not pattern or pattern.match(obj_name):
                response.append(obj_name)

        self.logger.info(f"files successfully listed from object store {self.object_store_name}")
        return response

    async def get(self, key: str) -> tuple[Optional[bytes], bool]:
        if not self.object_store:
            self.logger.warning(UNDEFINED_OBJECT_STORE)
            raise UndefinedObjectStoreError

        try:
            response = await self.object_store.get(key)
            self.logger.info(f"file {key} successfully retrieved from object store {self.object_store_name}")
            return response.data, True
        except ObjectNotFoundError as e:
            self.logger.debug(f"file {key} not found in object store {self.object_store_name}: {e}")
            return None, False
        except Exception as e:
            self.logger.warning(f"failed getting file {key} from object store {self.object_store_name}: {e}")
            raise FailedGettingFileError(key=key, error=e)

    async def save(self, key: str, payload: bytes) -> None:
        if not self.object_store:
            self.logger.warning(UNDEFINED_OBJECT_STORE)
            raise UndefinedObjectStoreError

        try:
            await self.object_store.put(key, payload)
            self.logger.info(f"file {key} successfully saved in object store {self.object_store_name}")
        except Exception as e:
            self.logger.warning(f"failed saving file {key} in object store {self.object_store_name}: {e}")
            raise FailedSavingFileError(key=key, error=e)

    async def delete(self, key: str) -> bool:
        if not self.object_store:
            self.logger.warning(UNDEFINED_OBJECT_STORE)
            raise UndefinedObjectStoreError

        try:
            info_ = await self.object_store.delete(key)
            return info_.info.deleted if info_.info.deleted else False
        except ObjectNotFoundError as e:
            self.logger.debug(f"file {key} not found in object store {self.object_store_name}: {e}")
            return False
        except Exception as e:
            self.logger.warning(f"failed deleting file {key} from object store {self.object_store_name}: {e}")
            raise FailedDeletingFileError(key=key, error=e)

    async def purge(self, regexp: Optional[str] = None) -> None:
        if not self.object_store:
            self.logger.warning(UNDEFINED_OBJECT_STORE)
            raise UndefinedObjectStoreError

        pattern = None
        if regexp:
            try:
                pattern = re.compile(regexp)
            except Exception as e:
                self.logger.warning(f"failed compiling regexp {regexp}: {e}")
                raise FailedCompilingRegexpError(error=e)

        object_names = await self.list()
        deleted = 0
        for name in object_names:
            if not pattern or pattern.match(name):
                self.logger.info(f"deleting file {name} from object store {self.object_store_name}...")

                try:
                    info_ = await self.object_store.delete(name)
                    if info_.info.deleted:
                        deleted += 1
                        self.logger.info(f"file {name} successfully deleted from object store {self.object_store_name}")
                except Exception as e:
                    self.logger.warning(f"failed deleting file {name} from object store {self.object_store_name}: {e}")
                    raise FailedPurgingFilesError(error=e)

        self.logger.info(f"{deleted} files successfully purged from object store {self.object_store_name}")
